### PR TITLE
Make boot_read_swap_state declaration public

### PR DESCRIPTION
**Pull request by de-nordic**
The PR contains series of commits that:
 - make `boot_read_swap_state` public and available from bootutil_public.h;
 - change `fixup_revert` to use provided flash area object pointer to obtain swap state by call to `boot_read_swap_state`, instead of previously used `boot_read_swap_state_by_id`; this makes ID param (`sec_id`) no longer needed;
 - makes `boot_swap_sector` call `boot_read_swap_state` on already opened scratch flash area, instead of using `boot_read_swap_state_by_id`, with the area ID, which caused double open on the flash area;
 - changes logic of `boot_set_pending_multi` by moving `flash_area_open` to the top of the function, making that the only place where flash area object is obtained in the function; switches the function to use `boot_read_swap_state`;
 - switches `boot_set_confirmed_multi` to use `boot_read_swap_state`;

### CODE REVIEW

**Comment by de-nordic** [comment]
@Sherryzhang2 can you look at the changes too?

File: boot/bootutil/src/bootutil_public.c [hunk 1]
```diff
@@ -524,26 +523,24 @@ boot_set_pending_multi(int image_index, int permanent)
...
 
     default:
         assert(0);
-        return BOOT_EBADIMAGE;
+        rc = BOOT_EBADIMAGE;
     }
+
+done:
+    flash_area_close(fap);
+    return 0;
```
Comment (utzig) [inline boot/bootutil/src/bootutil_public.c:543]: Should probably return `rc`, no?

Comment (de-nordic) [reply boot/bootutil/src/bootutil_public.c:543]: (after lots of swearing to himself, he has finally admitted) Yes, you are right @utzig.

File: boot/bootutil/include/bootutil/bootutil_public.h [hunk 2]
```diff
@@ -231,6 +231,18 @@ int boot_read_image_ok(const struct flash_area *fap, uint8_t *image_ok);
 int
 boot_read_swap_state_by_id(int flash_area_id, struct boot_swap_state *state);
 
+/**
+ * @brief Read the image swap state
+ *
+ * @param flash_area fa pointer to flash_area object.
```
Comment (Sherryzhang2) [inline boot/bootutil/include/bootutil/bootutil_public.h:237]: Remove "flash_area"?

File: boot/bootutil/include/bootutil/bootutil_public.h [hunk 3]
```diff
@@ -231,6 +231,18 @@ int boot_read_image_ok(const struct flash_area *fap, uint8_t *image_ok);
...
+ * @param state Struture for holding swap state.
+ *
+ * @return 0 on success, nonzero errno code on fail.
```
Comment (Sherryzhang2) [inline boot/bootutil/include/bootutil/bootutil_public.h:240]: "errno" -> "error"?
"fail" -> "failure"?

File: boot/bootutil/include/bootutil/bootutil_public.h (hunk 2 above)
Comment (de-nordic) [reply boot/bootutil/include/bootutil/bootutil_public.h]: Done

File: boot/bootutil/include/bootutil/bootutil_public.h (hunk 3 above)
Comment (de-nordic) [reply boot/bootutil/include/bootutil/bootutil_public.h]: Done

File: boot/bootutil/src/bootutil_public.c [hunk 4]
```diff
@@ -524,26 +523,24 @@ boot_set_pending_multi(int image_index, int permanent)
...
+        break;
 
     default:
         assert(0);
-        return BOOT_EBADIMAGE;
+        rc = BOOT_EBADIMAGE;
     }
+
+done:
+    flash_area_close(fap);
```
Comment (d3zd3z) [inline boot/bootutil/src/bootutil_public.c:542]: Admittedly, `flash_area_close` doesn't return an error in any current configuration. But we could do something like:
```c
    int rc2 = flash_area_close(fap);
    if (rc == 0) {
        rc = rc2;
    }
```

although, I'm not sure this is really necessary.

**Review by d3zd3z — APPROVED** [review]
I'm ok with this, my comment is only a suggestion and doesn't have to be changed.

**Comment by utzig** [comment]
@de-nordic Please squash all commits, reading one by one does not make a lot of sense.

File: boot/bootutil/src/bootutil_public.c (hunk 4 above)
Comment (de-nordic) [reply boot/bootutil/src/bootutil_public.c:542]: Unfortunately the original author, of the flash_area_close, did not predict any return value from the close function, so it has void type. I was considering changing that in Zephyr, but dropped the idea.

**Comment by de-nordic** [comment]
> 
> 
> @de-nordic Please squash all commits, reading one by one does not make a lot of sense.

OK, I have squashed to 3:
 - the one that published boot_read_swap_state;
 - the one that changes revert_fixup signature;
 - the one that applies boot_read_swap_state where possible;

**Review by utzig — APPROVED** [review]
LGTM. The last commit looks weird with two subjects and signed-off-by lines; otherwise nice refactoring.

**Comment by de-nordic** [comment]
> 
> 
> LGTM. The last commit looks weird with two subjects and signed-off-by lines; otherwise nice refactoring.

Sorry for that, fixed.

**Pull request merged**